subprojects/snapshots/src/main/java/org/gradle/internal/vfs/VirtualFileSystem.java

```diff
@@ -16,21 +16,33 @@
 
 package org.gradle.internal.vfs;
 
-import org.gradle.internal.snapshot.SnapshotHierarchy;
+import org.gradle.internal.snapshot.CompleteFileSystemLocationSnapshot;
+import org.gradle.internal.snapshot.MetadataSnapshot;
+
+import java.util.Optional;
 
 public interface VirtualFileSystem {
-    SnapshotHierarchy getRoot();
-    void update(UpdateFunction updateFunction);
 
     /**
-     * Updates the snapshot hierarchy, passing a {@link SnapshotHierarchy.NodeDiffListener} to the calls on {@link SnapshotHierarchy}.
+     * Returns the complete snapshot stored at the absolute path.
+     */
+    Optional<CompleteFileSystemLocationSnapshot> getSnapshot(String absolutePath);
+
+    /**
+     * Returns the complete snapshot stored at the absolute path.
+     */
+    Optional<MetadataSnapshot> getMetadata(String absolutePath);
+
+    /**
+     * Returns a hierarchy augmented by the information of the snapshot at the absolute path.
+     */
+    void store(String absolutePath, MetadataSnapshot snapshot);
```

Comment:
It looks like currently we are always calling this with a `CompleteFileSystemLocationSnapshot` that knows its own absolute path. Why do we only take a `MetadataSnapshot` here?

Reply:
I only allow `CompleteFileSystemLocationSnapshot` in the `VirtualFileSystem` interface now. Let's add support for `MetadataSnapshot` if we need it.